src/modules/tools/zprobe/CalibrationStrategy.cpp

```diff
@@ -0,0 +1,597 @@
+            lambda *= 3;
+            if (lambda > 1e10) {
+                stream->printf("Max lambda\n");
+                return false;
+            }
+            THEKERNEL->call_event(ON_IDLE);
+            goto again; // adjusted lambda, no need to recompute JTJ, JTr
+        } else {
+            float improvement = (1 - new_error / last_error);
+            stream->printf(" (%.3f%% improvement)\n", 100*improvement);
+            last_error = new_error;
+            save();
+            print_state();
+
+            if (improvement < 0.0001 ||
+                sq_sum_delta < 0.0001 * 0.0001) {
+                // improvement < 0.01 % or < 0.1 µm adjustment
+                stream->printf("Done\n");
+                return true;
+            }
+
+            // reduce damping
+            lambda /= 2;
+            if (lambda < 1e-7) lambda = 1e-7; // Reasonable lower limit for float precision
+        }
+    }
+    if (last_error < 0.001) return true; // We are within 1 µm, let's call that a success
+
+    stream->printf("Failed to converge after %d iterations\n", max_iterations);
+    return false;
+}
+
+
+int endstop_parameter_index(char parameter) {
+    switch(parameter) {
+    case 'X': return 0;
+    case 'Y': return 1;
+    case 'Z': return 2;
+    default: return -1;
+    }
+}
+
+std::function<void(float[3])> compute_rotation_transform(float p, float q, float offset, float inverse_r[3]) {
+    // setup rotation matrix
+    Vector3 x(1,0,0),
+            y(0,1,0),
+            z(p,q,sqrt(1-p*p-q*q));
+    // orthogonalize
+    y = z.cross(x).unit();
+    x = y.cross(z); // y and z are now orthogonal unit vectors
+
+    // store last row of R
+    inverse_r[0] = x[2];
+    inverse_r[1] = y[2];
+    inverse_r[2] = z[2];
+
+    return [x,y,z,offset](float p[3]) {
+        Vector3 q(p[0],p[1],p[2]);
+        // Note: multiply by R^T
+        p[0] = q.dot(x);
+        p[1] = q.dot(y);
+        p[2] = q.dot(z) + offset;
+    };
+}
+
+
+bool CalibrationStrategy::set_parameter(char parameter, float value) {
+    if (parameter == 'U' || parameter == 'V' || parameter == 'W') {
+        if (parameter == 'U') plane_u = value;
+        if (parameter == 'V') plane_v = value;
+        if (parameter == 'W') plane_offset = value;
+
+        update_compensation_transformation();
+        return true;
+    }
+    int endstop = endstop_parameter_index(parameter);
+    if (endstop >= 0) {
+        void *returned_data;
+
+        if (PublicData::get_value(endstops_checksum, trim_checksum, &returned_data)) {
+            // refresh cache
+            float *rtrim = static_cast<float *>(returned_data);
+            trim[0] = rtrim[0]; trim[1] = rtrim[1]; trim[2] = rtrim[2];
+
+            trim[endstop] = value;
+            return PublicData::set_value( endstops_checksum, trim_checksum, trim);
+        }
+        return false;
+    } else {
+        BaseSolution::arm_options_t options;
+        if(THEKERNEL->robot->arm_solution->get_optional(options)) {
```

Comment:
Thanks, I had assumed set_optional would replace all options by the ones in the arm_options_t.